lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -11715,7 +11715,7 @@ void Emit(ParseNode* pnode, ByteCodeGenerator* byteCodeGenerator, FuncInfo* func
         }
         else
         {
-            byteCodeGenerator->Writer()->PatchableProperty(Js::OpCode::LdFld, pnode->location, callObjLocation, cacheId, isConstructorCall);
+            byteCodeGenerator->Writer()->PatchableProperty(pnode->isUsed ? Js::OpCode::LdFld_ReuseLoc : Js::OpCode::LdFld, pnode->location, callObjLocation, cacheId, isConstructorCall);
```

Comment:
I think _ReuseLoc is only wanted here if we're in an optional chain.

If we're in an optional chain then we've written undefined to the location AND the result of this operation needs to overwrite it so that some other operation can observe EITHER undefined or this op.

If we're in a non-optional chain, then the location won't have been pre-written to with any relevant value.

Some of the test fails say a bytecode regen is needed and I think that stems from some of the self-hosted JS methods having followed this path and INCORRECTLY having swapped to LdFld_ReuseLoc.

Reply:
Another thought on this actually. It may be better if we didn't have to use reuseLoc throughout the chain. In an optional chain there are two possibilities:
1. Run to the end and get the result
2. Shortcut at some point and have undefined

reuseLoc is used when writing to a location we've written to before with it mattering that it's the same location e.g. we write under some condition AND the value is checked regardless of the condition.

The operations throughout the chain don't need to reuseLoc as there are only 2 loads that can write a value that will actually be read from the Location we're writing the chain to: a) the final result and b) undefined.

So we ought to only need to use _reuseLoc once at most (if we do the Ldundef first) OR perhaps we can avoid it altogether, IF you write the LdUndef NOT at the start BUT at the end.

Reply:
Test for a9a1c7019a0e21a4a847deb23a4ed418a87b8d9c are failing because I did not regenerate bytecode

Reply:
> Test for [a9a1c70](https://github.com/chakra-core/ChakraCore/commit/a9a1c7019a0e21a4a847deb23a4ed418a87b8d9c) are failing because I did not regenerate bytecode

Yes, but it only wanted a byte code regen because some of the property access in the builtin JS methods were being marked as _reuseLoc (when they actually shouldn't be)

Reply:
Bytecode Regen should only be needed if:
a) you change the byte code implementation for existing code OR
b) you cause byte code ops to be renumbered (e.g. by adding a new operation so if you do add LdUndef_reuseLoc it probably will need it)

Reply:
I'm now using `_ReuseLoc` for the `undefined` and chain-result value.
Instead of adding a new opcode, I use `Ld_A_ReuseLoc result undefined`